File: models/Balance/Operation.cs
using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using ReconBank.Models.Users;

namespace ReconBank.Models.Balance
{
    public class Operation
    {
        public Guid Id { get; set; }

        [NotMapped]
        public OperationType Type
        {
            get
            {
                if (this.DestinationId.HasValue)
                {
                    return OperationType.TRANSFER;
                }

                return this.AmountInCents == Math.Abs(this.AmountInCents)
                    ? OperationType.INCOMING
                    : OperationType.OUTGOING;
            }
        }

        [Required]
        public Guid OriginId { get; set; }

        public virtual User Origin { get; set; }

        public Guid? DestinationId { get; set; }

        public virtual User Destination { get; set; }

        [Required]
        public bool OwnTitularity { get; set; }

        [Required]
        public int AmountInCents { get; set; }

        public DateTime Timestamp { get; set; }
    }
}

File: front_user/ClientApp/src/models/operation.ts
import * as yup from "yup";

import { OperationType } from "./operationType";
import { User } from "./user";

export interface Operation {
  id?: string;
  type: OperationType;
  originId?: string;
  origin?: User | null;
  destinationId?: string | null;
  destination?: User | null;
  amountInCents: number;
  ownTitularity: boolean;
  timestamp: Date;
}

export const operationSchema = yup.object().shape({
  type: yup.number().required().positive().integer(),
  destinationId: yup.string().when("type", {
    is: (val: OperationType) => val === OperationType.TRANSFER,
    then: yup
      .string()
      .nullable()
      .required("Destination user is required when making a transfer"),
    otherwise: yup.string().nullable(),
  }),
  amountInCents: yup
    .number()
    .required()
    .positive("Amount in cents must be at least 0,01")
    .integer(),
  ownTitularity: yup.bool().required(),
});
